### ISSUE TITLE
`self_named_module_files` does not find self named module files

### ISSUE THREAD

**Issue by nerosnm** [rust-lang/rust-clippy#8094]
### Summary

The `self_named_module_files` lint fails to trigger an error in a crate with this file structure:

```
.
├── Cargo.lock
├── Cargo.toml
└── src
   ├── lib.rs
   ├── outer
   │  └── inner.rs
   └── outer.rs
```

I initially thought the issue might be related to the inner module being annotated with `#[cfg(test)]`, but that's not needed to cause the false negative.

### Lint Name

self_named_module_files

### Reproducer

I tried this code:

```rust
// src/lib.rs
#![deny(clippy::self_named_module_files)]
pub mod outer;
```

```rust
// src/outer.rs
pub mod inner;
```

```rust
// src/outer/inner.rs
pub struct Foo;
```

I expected to see this happen: The `self_named_module_files` lint causes an error when `cargo clippy` is run.

Instead, this happened: No lints caused any errors.

### Version

```text
rustc 1.57.0 (f1edd0429 2021-11-29)
binary: rustc
commit-hash: f1edd0429582dd29cccacaf50fd134b05593bd9c
commit-date: 2021-11-29
host: x86_64-unknown-linux-gnu
release: 1.57.0
LLVM version: 13.0.0

The same issue also occurs on the latest nightly at the time of filing this issue:

rustc 1.59.0-nightly (0b6f079e4 2021-12-07)
binary: rustc
commit-hash: 0b6f079e4987ded15c13a15b734e7cfb8176839f
commit-date: 2021-12-07
host: x86_64-unknown-linux-gnu
release: 1.59.0-nightly
LLVM version: 13.0.0
```

**Comment by nerosnm** [comment]
Here's the minimal reproduction described in the issue: https://github.com/nerosnm/clippy-mod-file-false-negative.

**Comment by nerosnm** [comment]
I should add that it doesn't fail in every case, for me. In the private repo where I initially noticed this, which is a workspace with 4 members, only one instance of a self named module file wasn't found, while 3 or so were. _But_, the weird thing about the ones that were found was that they were only found by dependent crates with the lint set to `deny`, and not the crate they actually live in. Perhaps that's the root of the issue here.

**Comment by DevinR528** [comment]
Lint author here (although I'm not in the file history)

So this case is just not covered by the lint, and I'm not sure it could be the current way it's checked. If the `deny` was put at the top level it would trigger for the `outer` mod because it has a matching folder but because the lint only checks for "self named mod" files when it finds a correspondingly named folder.

The `inner` name is never checked since it's not in `folder_segments`
https://github.com/rust-lang/rust-clippy/blob/master/clippy_lints/src/module_style.rs#L116

**Comment by nerosnm** [comment]
I'm not sure I understand - but it might just be that I haven't looked carefully enough at the implementation of the lint. You say "if the `deny` was put at the top level it would trigger", but the lint *is* at the top level in my minimal repro - it's enabled using an outer attribute.

To me, it seems like this case is the simple general case for this lint rather than some kind of weird edge case.

**Comment by DevinR528** [comment]
> You say "if the deny was put at the top level it would trigger", but the lint is at the top level in my minimal repro - it's enabled using an outer attribute.

Sorry, I'm not sure how I missed `// src/lib.rs` at the top of the first example my bad :facepalm: . Also the reason for your confusion may be that I'm wrong about the lint triggering for `inner.rs` it should trigger for `outer/` since what the lint checks is that `outer/` has no file `src/outer/mod.rs` then the `clippy::self_named_module_files` lint triggers.

**Comment by ivan-aksamentov** [comment]
@DevinR528 This rule emits false positives for files in `src/bin/`. If a bin crate "foo" has `src/bin/foo.rs` or `src/bin/bar.rs` it should not cause any warnings, because this is totally normal.

And no, `mod.rs` files are not "required" in any case. I would soften the language of the warning, it is currently too aggressive and opinionated.

**Comment by Nemo157** [comment]
> And no, `mod.rs` files are not "required" in any case. I would soften the language of the warning, it is currently too aggressive, opinionated and overall too short and confusing. This is especially important for newcomers, who may take this rule as a call to action.

I thought the point of this lint is that by activating it _you_ require usage of `mod.rs` files. That at least is my intended usecase, I find `mod.rs` files much easier to work with and want to ensure that both contributors and myself do not accidentally try to use the "new" module style.

**Comment by ivan-aksamentov** [comment]
@Nemo157 Surely. I use `mod.rs` style everywhere too.

But if you get this warning, and it's not you who enabled it, the message does not make much sense. It has only 4 words and aggressively imposes something, without teaching why, and what to do with all that. This is the most frustrating type of lints and compiler errors. All combined with the fact that Rust module system is not a strike of a genius either, plus its docs are pretty mediocre and lots of outdated stuff on the internet.

I can see how an intern sees this warning, renames the `foo.rs` into `mod.rs`, as "required", and then spends a day figuring out why `bar.rs` is no longer showing up in the project.

I would write something like: "This project is using `mod.rs` convention, but `<dir>/mod.rs` is not found and `<dir>/<file>` is found instead. Consider creating `<dir>/mod.rs` containing `mod foo; mod bar;`." 

This would give some words to google, some context and a proposed action. Perhaps native English speakers can suggest a shorter alternative.

An autofix would be fantastic.

And false positives need to be fixed of course. `clippy::restriction` is not intended to contain buggy rules. That's for `nursery`.
